core/primitives/src/stateless_validation.rs

```diff
@@ -24,16 +24,29 @@ type SignatureDifferentiator = String;
 #[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize)]
 pub struct EncodedChunkStateWitness(Box<[u8]>);
 
+pub type ChunkStateWitnessSize = usize;
+
 impl EncodedChunkStateWitness {
-    pub fn encode(witness: &ChunkStateWitness) -> Self {
-        Self(borsh::to_vec(witness).expect("borsh serialization should not fail").into())
+    /// Borsh-serialize and compress state witness.
+    /// Returns encoded witness along with the raw (uncompressed) witness size.
+    pub fn encode(witness: &ChunkStateWitness) -> std::io::Result<(Self, ChunkStateWitnessSize)> {
+        const STATE_WITNESS_COMPRESSION_LEVEL: i32 = 3;
+        let borsh_bytes = borsh::to_vec(witness)?;
+        Ok((
+            Self(zstd::encode_all(borsh_bytes.as_slice(), STATE_WITNESS_COMPRESSION_LEVEL)?.into()),
+            borsh_bytes.len(),
+        ))
     }
 
-    pub fn decode(&self) -> Result<ChunkStateWitness, std::io::Error> {
-        ChunkStateWitness::try_from_slice(&self.0)
+    /// Decompress and borsh-deserialize encoded witness bytes.
+    /// Returns decoded witness along with the raw (uncompressed) witness size.
+    pub fn decode(&self) -> std::io::Result<(ChunkStateWitness, ChunkStateWitnessSize)> {
+        let borsh_bytes = zstd::decode_all(self.0.as_ref())?;
+        let witness = ChunkStateWitness::try_from_slice(&borsh_bytes)?;
+        Ok((witness, borsh_bytes.len()))
     }
```

Comment:
Is it possible to construct [Zip bombs](https://en.wikipedia.org/wiki/Zip_bomb) with zstd? I'm worried that someone might be able to send something that would decompress to multiple terabytes in size, which would crash the node.

Reply:
Interesting.... I'm working on dividing witness into parts and sending over network and I would probably require some sort of header + body type setup. We can consider using the header for basic validation and only then using the body for unzip + full validation.

Reply:
@jancionear this is a very good point, thanks a lot for highlighting that! I will look into this.

Reply:
addressed in 9ad0851e552de516dcd63c5442b4fbc371fedf59